src/generators/non_gradient_based/feature_tweak/feature_tweak.jl

```diff
@@ -7,73 +7,77 @@ mutable struct FeatureTweakGenerator <: AbstractNonGradientBasedGenerator
-- `x::AbstractArray`: The factual instance.
-- `target::RawTargetType`: The target class.
+- `ce::AbstractCounterfactualExplanation`: The counterfactual explanation object.
 
 # Returns
-- `x_out::AbstractArray`: The counterfactual instance.
+- `ce::AbstractCounterfactualExplanation`: The counterfactual explanation object.
 
 # Example
-x = feature_tweaking(generator, M, x, target) # returns a counterfactual instance of `x` based on the ensemble of classifiers provided
+ce = feature_tweaking(ce) # returns a counterfactual inside the ce.s′ field based on the ensemble of classifiers provided
 """
-function feature_tweaking(
-    generator::FeatureTweakGenerator,
-    M::Models.TreeModel,
-    x::AbstractArray,
-    target::RawTargetType,
-)
-    if Models.predict_label(M, x) == target
+function feature_tweaking(ce::AbstractCounterfactualExplanation)
+    @assert isa(ce.generator, Generators.FeatureTweakGenerator) "The feature tweak algorithm can only be applied using the feature tweak generator"
+    @assert isa(ce.M, Models.TreeModel) "The `FeatureTweakGenerator` currently only supports tree models. The counterfactual search will be terminated."
+
+    if Models.predict_label(ce.M, ce.x) == ce.target
         return ce
     end
 
-    x_out = deepcopy(x)
     delta = 10^3
-    ensemble_prediction = Models.predict_label(M, x)[1]
+    ensemble_prediction = Models.predict_label(ce.M, ce.x)[1]
 
-    for classifier in Models.get_individual_classifiers(M)
-        if ensemble_prediction != target
+    for classifier in Models.get_individual_classifiers(ce.M)
+        if ensemble_prediction != ce.target
             y_levels = MLJBase.classes(
-                MLJBase.predict(M.model, DataFrames.DataFrame(x', :auto))
+                MLJBase.predict(ce.M.model, DataFrames.DataFrame(ce.x', :auto))
             )
-            paths = search_path(classifier, y_levels, target)
+            paths = search_path(classifier, y_levels, ce.target)
+
             for key in keys(paths)
                 path = paths[key]
-                es_instance = esatisfactory_instance(generator, x, path)
-                if target .== Models.predict_label(M, es_instance)[1]
-                    if LinearAlgebra.norm(x - es_instance) < delta
-                        x_out = es_instance
-                        delta = LinearAlgebra.norm(x - es_instance)
+                es_instance = esatisfactory_instance(ce.generator, ce.x, path)
+                if ce.target .== Models.predict_label(ce.M, es_instance)[1]
+                    s′_old = ce.s′
+                    ce.s′ = reshape(es_instance, :, 1)
+                    if ce.generator.penalty(ce) < delta
+                        delta = ce.generator.penalty(ce)
+                    else
+                        ce.s′ = s′_old
                     end
                 end
             end
         end
     end
-    return x_out
+
+    return ce
```

Comment:
Since this just returns its only positional argument, should we either 1) turn this into an in-place operation `feature_tweak!(ce)` or 2) keep this, but also add an in-place version `feature_tweak!(ce)` version? 

I think I have a slight preference for option (1), because this would be a non-breaking change and it just makes sense to use the in-place version here.

Reply:
Yes, I think it makes the most sense to just turn this into an in-place operation. Will do so in a moment.